File: 2026/January/22/hammingdistance.cpp
// LeetCode 461: Hamming Distance
// Method: XOR + Brian Kernighan’s Bit Counting
// Time Complexity: O(number of set bits)
// Space Complexity: O(1)

#include <iostream>
using namespace std;

class Solution {
public:
    int hammingDistance(int x, int y) {
        int n = x ^ y;
        int count = 0;
        while (n) {
            n &= (n - 1);
            count++;
        }
        return count;
    }
};

File: 2026/January/22/hammingdistance.py
# LeetCode 461: Hamming Distance
# Explanation:
# The Hamming distance between two integers is the number of positions
# at which the corresponding bits are different.
#
# Method 1: XOR + Bit Counting (Optimal)
# - XOR x and y → differing bits become 1
# - Count number of 1s in the result
#
# Time Complexity: O(1) (bounded by number of bits, max 32)
# Space Complexity: O(1)

class Solution:
    def hammingDistance(self, x: int, y: int) -> int:
        return bin(x ^ y).count("1")
